symengine/sets.cpp

```diff
@@ -804,4 +911,46 @@ RCP<const Set> set_complement(const RCP<const Set> &universe,
     return container->set_complement(universe);
 }
 
+RCP<const Set> conditionset(const RCP<const Basic> &sym,
+                            const RCP<const Boolean> &condition)
+{
+    if (eq(*condition, *boolean(false))) {
+        return emptyset();
+    } else if (eq(*condition, *boolean(true))) {
+        return universalset();
+    }
+    if (is_a<And>(*condition)) {
+        auto cont = down_cast<const And &>(*condition).get_container();
+        set_boolean newcont;
+        set_set temp;
+        for (auto it = cont.begin(); it != cont.end(); it++) {
+            if (is_a<Contains>(**it)
+                and eq(*down_cast<const Contains &>(**it).get_expr(), *sym)
+                and is_a<FiniteSet>(
+                        *down_cast<const Contains &>(**it).get_set())) {
+                auto fset = down_cast<const Contains &>(**it).get_set();
+                auto fcont
+                    = down_cast<const FiniteSet &>(*fset).get_container();
+                // use the result of simplification done in `logical_and()`
+                if (has_all_numbers(fcont)) {
+                    temp.insert(fset);
+                } else {
+                    newcont.insert(*it);
+                }
+            } else {
+                newcont.insert(*it);
+            }
+        }
+        if (not temp.empty()) {
+            return SymEngine::set_union(
+                {SymEngine::set_union(temp),
+                 conditionset(sym, logical_and(newcont))});
+        }
```

Comment:
SymPy doesn't do the simplification either, `ConditionSet(x, And(FiniteSet(-1, 0, 2).contains(x), x > 0), S.Reals)`